src/round.rs

```diff
@@ -0,0 +1,327 @@
+// This file is part of finality-afg.
+
+// finality-afg is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// finality-afg is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with finality-afg. If not, see <http://www.gnu.org/licenses/>.
+
+//! Logic for a single round of AfG.
+
+use vote_graph::VoteGraph;
+
+use std::collections::hash_map::{HashMap, Entry};
+use std::hash::Hash;
+use std::ops::AddAssign;
+
+use super::{Equivocation, Prevote, Precommit, Chain};
+
+#[derive(Hash, Eq, PartialEq)]
+struct Address;
+
+#[derive(Default, Debug, Clone)]
+struct VoteCount {
+	prevote: usize,
+	precommit: usize,
+}
+
+impl AddAssign for VoteCount {
+	fn add_assign(&mut self, rhs: VoteCount) {
+		self.prevote += rhs.prevote;
+		self.precommit += rhs.precommit;
+	}
+}
+
+struct VoteTracker<Id: Hash + Eq, Vote, Signature> {
+	votes: HashMap<Id, (Vote, Signature)>,
+	current_weight: usize,
+}
+
+impl<Id: Hash + Eq + Clone, Vote: Clone + Eq, Signature: Clone> VoteTracker<Id, Vote, Signature> {
+	fn new() -> Self {
+		VoteTracker {
+			votes: HashMap::new(),
+			current_weight: 0,
+		}
+	}
+
+	// track a vote. if the vote is an equivocation, returns a proof-of-equivocation and
+	// otherwise notes the current amount of weight on the tracked vote-set.
+	//
+	// since this struct doesn't track the round number of votes, that must be set
+	// by the caller.
+	fn add_vote(&mut self, id: Id, vote: Vote, signature: Signature, weight: usize)
+		-> Result<(), Equivocation<Id, Vote, Signature>>
+	{
+		match self.votes.entry(id.clone()) {
+			Entry::Vacant(mut vacant) => {
+				vacant.insert((vote, signature));
+			}
+			Entry::Occupied(mut occupied) => {
+				if occupied.get().0 != vote {
+					return Err(Equivocation {
+						round_number: 0,
+						identity: id,
+						first: occupied.get().clone(),
+						second: (vote, signature),
+					})
+				}
+			}
+		}
+
+		self.current_weight += weight;
+
+		Ok(())
+	}
+}
+
+/// Parameters for starting a round.
+pub struct RoundParams<Id: Hash + Eq, H> {
+	/// The round number for votes.
+	pub round_number: u64,
+	/// Actors and weights in the round.
+	pub voters: HashMap<Id, usize>,
+	/// The base block to build on.
+	pub base: (H, usize),
+}
+
+#[derive(Debug)]
+pub enum Error<Id, H, S> {
+	PrevoteEquivocation(Equivocation<Id, Prevote<H>, S>),
+	PrecommitEquivocation(Equivocation<Id, Precommit<H>, S>),
+	Chain(::Error),
+}
+
+/// Stores data for a round.
+pub struct Round<Id: Hash + Eq, H: Hash + Eq, Signature> {
+	graph: VoteGraph<H, VoteCount>,
```

Comment:
can't hurt, but the fields aren't pub so my convention is not to add doc comments. in this case I will add a few because the code isn't easy to understand without